extensions/wikia/Captcha/Module/BaseCaptcha.class.php

```diff
@@ -0,0 +1,235 @@
+		return [
+			'question' => $test,
+			'answer' => $answer
+		];
+	}
+
+	public function addCaptchaAPI( &$resultArr ) {
+		$captcha = $this->getCaptcha();
+		$index = $this->storeCaptcha( $captcha );
+		$resultArr['captcha']['type'] = 'simple';
+		$resultArr['captcha']['mime'] = 'text/plain';
+		$resultArr['captcha']['id'] = $index;
+		$resultArr['captcha']['question'] = $captcha['question'];
+	}
+
+	/**
+	 * Insert a captcha prompt into the edit form.
+	 * This sample implementation generates a simple arithmetic operation;
+	 * it would be easy to defeat by machine.
+	 *
+	 * Override this!
+	 *
+	 * @return string HTML
+	 */
+	public function getForm() {
+		$captcha = $this->getCaptcha();
+		$index = $this->storeCaptcha( $captcha );
+
+		return "<p><label for=\"wpCaptchaWord\">{$captcha['question']}</label> = " .
+		\Xml::element( 'input', [
+			'name' => 'wpCaptchaWord',
+			'id'   => 'wpCaptchaWord',
+			'tabindex' => 1,
+		] ) . // tab in before the edit textarea
+		"</p>\n" .
+		\Xml::element( 'input', [
+			'type'  => 'hidden',
+			'name'  => 'wpCaptchaId',
+			'id'    => 'wpCaptchaId',
+			'value' => $index,
+		] );
+	}
+
+	/**
+	 * Show a message asking the user to enter a captcha on edit
+	 * The result will be treated as wiki text
+	 *
+	 * @param string $action Action being performed
+	 *
+	 * @return string
+	 */
+	public function getMessage( $action ) {
+		$name = 'captcha-' . $action;
+		$text = wfMsg( $name );
+		# Obtain a more tailored message, if possible, otherwise, fall back to
+		# the default for edits
+		return wfEmptyMsg( $name, $text ) ? wfMsg( 'captcha-edit' ) : $text;
+	}
+
+	/**
+	 * Check if the submitted form matches the captcha session data provided
+	 * by the plugin when the form was generated.
+	 *
+	 * Override this!
+	 *
+	 * @param string $answer
+	 * @param array $info
+	 *
+	 * @return bool
+	 */
+	public function keyMatch( $answer, $info ) {
+		return $answer == $info['answer'];
+	}
+
+	/**
+	 * @param \ApiBase $module
+	 * @param array $params
+	 *
+	 * @return bool
+	 */
+	public function APIGetAllowedParams( &$module, &$params ) {
+		if ( !$module instanceof \ApiEditPage ) {
+			return true;
+		}
+		$params['captchaword'] = null;
+		$params['captchaid'] = null;
+
+		return true;
+	}
+
+	/**
+	 * @param \ApiBase $module
+	 * @param array $desc
+	 * @return bool
+	 */
+	public function APIGetParamDescription( &$module, &$desc ) {
+		if ( !$module instanceof \ApiEditPage ) {
+			return true;
+		}
+		$desc['captchaid'] = 'CAPTCHA ID from previous request';
+		$desc['captchaword'] = 'Answer to the CAPTCHA';
+
+		return true;
+	}
+
+	/**
+	 * Given a required captcha run, test form input for correct
+	 * input on the open session.
+	 *
+	 * @return bool if passed, false if failed or new session
+	 */
+	public function passCaptcha() {
+		$info = $this->retrieveCaptcha();
+		if ( $info ) {
+			if ( $this->keyMatch( \F::app()->wg->Request->getVal( 'wpCaptchaWord' ), $info ) ) {
+				$this->log( "passed" );
```

Comment:
I noticed most these strings passed to the `log` message have a newline at the end but these here don't. Do we need to include that newline? Seems like it would be easier if we left it off so we don't have to remember to add it for all of these. It shouldn't affect how Kibana displays them, right? If we do need those newlines, then we should add them to these log calls for consistency.